### PR TITLE
feat(ibkr): group partial-fill executions into single trades

### PR DESCRIPTION
**Pull request by lucasfarre**
## Summary

IBKR Flex Query emits one `<Trade levelOfDetail="EXECUTION">` per fill, so an order that fills in 3 partial executions yields 3 separate Trade rows in our parser. Brokers that report fiscally to Hacienda (and competitors like Autodeclaro / Taxdown) aggregate at the **order level**, so the operations annex was showing many more lines than the user sees on their own broker statement — confusing to reconcile, harder to audit.

This collapses same-order EXECUTION rows by `ibOrderID` + `tradeDate` into a single synthetic Trade. **Tax math is unchanged**: the FIFO engine consumes the exact same totals because the merge preserves the identity

```
tradeMoney = quantity × tradePrice × multiplier
```

— summed quantities and money divide back to the volume-weighted average price.

### What changed

**Parser (`src/parsers/ibkr.ts`)**

- `mergeExecutionsByOrder()` groups EXECUTION rows by `ibOrderID|tradeDate` and aggregates quantity, tradeMoney, proceeds, cost, commission, taxes and `fifoPnlRealized` via `Decimal`. `tradePrice` becomes the VWAP. `notes` are split-trimmed-deduplicated into a `Set` so per-fill markers like `Ep` / `P` survive.
- **GTC year-boundary safe**: the group key includes `tradeDate`, so a GTC order that fills on Dec 31 and Jan 2 splits into two rows landing in the correct tax year (Art. 14 LIRPF).
- Trades with empty `ibOrderID` pass through un-merged. The IBKR Flex schema marks `ibOrderID` as optional (csingley/ibflex `Types.py`: `ibOrderID: str | None = None`), and BookTrade / internal-transfer rows can legitimately omit it.
- New `parserMessages` info entry `parser.executions_merged` surfaces the count so the user knows what happened.

**ORDER-level duplicate guard**

Symmetric to the cash-transactions guard added in #194 (`isCashSummaryRow`): IBKR Flex Query allows multiple `levelOfDetail` selections on the Trades section. If a user enables both **Executions** (per-fill) and **Orders** (aggregated), each activity is emitted twice and the merge above would silently **double-count** quantity, money and commission because both row types share the same `ibOrderID`.

`filterDuplicateLevelOfDetail()` drops ORDER-level rows whenever EXECUTION rows are also present in the same FlexStatement. Pure ORDER-only exports remain valid input (the filter is a no-op). New `parserMessages` info entry `parser.order_level_duplicates` reports the count.

**Types (`src/types/ibkr.ts`)**

- New optional `Trade.ibOrderID` field (string). Backwards compatible.

### Legal basis

- **Art. 37.2 LIRPF** — FIFO over homogeneous securities. Merging at order level does not change which lots are consumed; the per-lot disposals fall out identically.
- **Art. 14 LIRPF** — devengo by trade date. The `tradeDate` component of the group key preserves per-day granularity required for year-boundary GTC orders.

## Test plan

- [x] `npm run typecheck` — clean
- [x] `npm test` — **1277 passed** (+7 new: 3-fill VWAP, GTC split, mixed single/multi-fill, empty `ibOrderID` passthrough, OPT `Ep` note preservation, ORDER+EXECUTION drop, ORDER-only no-op)
- [x] Sanity: re-running a real EXECUTION-only Flex Query (40 trade rows, 6 multi-fill orders) produces unchanged capital gains and the expected 6-orders collapse, with the `parser.executions_merged` info message reporting `12 ejecuciones parciales en 6 órdenes`.

## Summary by CodeRabbit

* **New Features**
  * Consolidate IBKR partial-fill executions sharing the same order ID and trade date into a single trade with VWAP-style pricing (with fallback), summed amounts, de-duplicated notes, and the order ID included in mapped output.

* **Bug Fixes**
  * Omit duplicate ORDER-level rows when corresponding EXECUTION-level rows exist.

* **Tests**
  * Added thorough tests for merging, duplicate filtering, VWAP/fallback behavior, and multiple edge cases; parser emits informational counts for merges and duplicates.

### CODE REVIEW

File: src/parsers/ibkr.ts [hunk 1]
```diff
@@ -182,13 +213,147 @@ function mapTrade(raw: Record<string, string>): Trade {
...
+ * trades only drop ORDER when EXECUTION is also present (an ORDER-only export
+ * remains valid input).
+ */
+function filterDuplicateLevelOfDetail(rawTrades: Record<string, string>[]): Record<string, string>[] {
+  const hasExecution = rawTrades.some(
+    (t) => (t.levelOfDetail ?? "").toUpperCase() === LEVEL_OF_DETAIL_EXECUTION,
+  );
+  if (!hasExecution) return rawTrades;
+  return rawTrades.filter((t) => (t.levelOfDetail ?? "").toUpperCase() !== LEVEL_OF_DETAIL_ORDER);
+}
```
[the remark this answers is not in the document]
Comment (lucasfarre) [reply src/parsers/ibkr.ts:391]: Done in 49e0d63. Filter now builds a Set of `ibOrderID`s seen at `EXECUTION` level and drops an `ORDER` row only when its own `ibOrderID` is in that set. `ORDER` rows with empty `ibOrderID` are always kept (BookTrade / internal-transfer edge cases). Pure ORDER-only export remains a no-op. Added a mixed-LOD test: ORD-X merges 2 EXECUTION fills and drops its matching ORDER aggregate, ORD-Y standalone ORDER is preserved.

**Comment by GeiserX** [comment]
Thanks for this @lucasfarre — really high-quality PR: the `ibOrderID|tradeDate` key, the GTC year-boundary handling, the per-`ibOrderID` ORDER/EXECUTION dedup (avoiding the global-flag trap that bit our FX engine), and the thorough tests are all spot-on. I ran it through a multi-perspective review (correctness, AEAT primary sources, and an adversarial edge-case pass). Summary: **approach is fiscally sound and the VWAP precision concern checks out — a few edge-case guards before merge.**

### ✅ Verified safe
- **Fiscal legitimacy** (Art. 37.2 FIFO + Art. 14.1.c devengo, BOE-A-2006-20764): merging same-day/same-order partial fills at VWAP is equivalent to per-fill for the IRPF result, and splitting across `tradeDate` (GTC) is *required*, not optional. Correct.
- **VWAP precision**: even though FIFO recomputes value from `tradePrice × quantity × multiplier` (it doesn't read the summed `tradeMoney`/`proceeds`), with our `Decimal` precision-20 the VWAP round-trip drifts ~1e-11 EUR — ~6 orders of magnitude below a cent, can't flip a `toFixed(2)`. The "tax math unchanged" claim holds.

### 🔧 Please address before merge

**1. (high) `tradeID: ""` on merged trades collides in the dedup validator.** `src/web/validation.ts:80` keys on `tradeID` and falls back to a composite only when falsy — so every merged trade hits the fallback and two distinct merged orders matching on date+symbol+netQty+VWAP+side would be flagged as a spurious "duplicate trades" warning on correct data. It also makes the `merge.ts:52` sort tiebreaker (`a.tradeID.localeCompare(b.tradeID)`) a no-op → non-deterministic display order (FIFO itself re-sorts independently at `fifo.ts:76`, so tax math is unaffected — display only).
   → Suggest a stable synthetic id: `tradeID: \`merged-${first.ibOrderID}-${first.tradeDate}\``.

**2. (high) No guard for a mixed `buySell` or `openCloseIndicator` group.** The merge keys on `ibOrderID|tradeDate` and takes `...first` for `buySell`/`openCloseIndicator`, but sums signed quantity and VWAPs `money.abs()/qtyAbs`. If one order has both BUY and SELL fills, or partially closes then opens (a flip), the VWAP is meaningless and the O/C is wrong — and `fifo.ts:80` branches on O/C, so it's tax-relevant when it happens.
   → Extend the group key to `ibOrderID|tradeDate|buySell|openCloseIndicator` (so heterogeneous fills stay separate), and key `filterDuplicateLevelOfDetail` on `ibOrderID|tradeDate` too (it currently matches by `ibOrderID` only, which could drop a GTC-id-reused ORDER row on a different date).

**3. (nice-to-have) Tests for the above**: a mixed-`buySell` group, a mixed-O/C group, and two distinct merged orders colliding in the validator. The current suite only covers homogeneous all-BUY/all-O groups, so these paths are uncovered.

None of this detracts from a genuinely excellent contribution — the common path is correct and well-guarded. Happy to help with any of these if useful. 🙏

**Comment by lucasfarre** [comment]
Thanks @GeiserX — all three addressed in 9554f7b (with the per-`ibOrderID` dedup tightening already in bb81c07).

**1. Validator collision on empty `tradeID`.** Merged trades now carry a stable synthetic id:

```
merged-${ibOrderID}-${tradeDate}-${buySell}-${openCloseIndicator}
```

Mirrors the full bucket key from fix #2, so two buckets under one `ibOrderID` (e.g. a position flip with multi-fill Open and multi-fill Close legs) cannot alias to the same `tradeID`. Greppable too. Fixes the spurious `validation.ts:80` duplicates and restores a deterministic tiebreaker on the `merge.ts` sort.

**2. Mixed `buySell` / `openCloseIndicator` in one `ibOrderID`.** Merge group key extended to `ibOrderID|tradeDate|buySell|openCloseIndicator`, so combo/spread legs and position flips stay in separate buckets — no more meaningless VWAP across opposite sides, and `fifo.ts:80` sees the correct O/C.

For the dedup side, `filterDuplicateLevelOfDetail` now keys on `(ibOrderID, tradeDate)` (bb81c07 had already moved it off the global flag; 9554f7b adds the date to the key). A GTC order whose `ibOrderID` carries across days no longer risks an ORDER row being silently dropped on a date with no matching EXECUTION.

**3. Tests.** Added in `tests/parsers/ibkr.test.ts`:
- BUY + SELL fills under one `ibOrderID` → two separate merged trades, signs preserved
- Open + Close fills under one `ibOrderID` → two buckets, O/C preserved
- Two distinct merged orders that collide on the validator's composite fallback → both carry unique non-empty `tradeID`s
- Mixed-LOD: ORDER row on a date with no matching EXECUTION survives the filter (covers the GTC-across-days case)
- Position flip with multi-fill Open + multi-fill Close → synthetic `tradeID` stays unique across the two same-`(ibOrderID, date)` buckets

Branch is `lucasfarre:feat/ibkr-execution-merge` at 9554f7b, ready for another look.

**Review by GeiserX — APPROVED** [review]
Re-reviewed at 9554f7b — all three resolved cleanly, thanks @lucasfarre. 🙌

**1. Synthetic tradeID** — `merged-${ibOrderID}-${tradeDate}-${buySell}-${openCloseIndicator}` mirrors the bucket key exactly, so it's unique per merged bucket; `validation.ts:80` now sees distinct non-empty ids (no false "duplicate trades") and the `merge.ts` sort tiebreaker is deterministic again. Verified the position-flip case keeps the O and C buckets on distinct ids.

**2. Heterogeneous fills** — bucket key `ibOrderID|tradeDate|buySell|openCloseIndicator` keeps a BUY-fill and SELL-fill (and Open vs Close) in separate buckets, so no VWAP across opposite sides and `fifo.ts` sees the correct O/C. The single-row `C;O` flip is its own singleton bucket and passes through untouched. `filterDuplicateLevelOfDetail` keyed on `(ibOrderID, tradeDate)` correctly preserves an ORDER row on a date with no matching EXECUTION (the GTC-across-days case).

**3. Tests** — all five scenarios present with meaningful numeric assertions (BUY+SELL split, Open+Close split, two distinct merged orders → unique ids, ORDER-survives-on-no-EXECUTION-date, position-flip multi-fill). 

Also re-confirmed: VWAP round-trip stays sub-1e-11 EUR (FIFO recomputes `tradePrice×qty×mult`, bucketing preserves each bucket's `Σmoney == Σ(qty×price×mult)` identity); splitting a flip into 2 buckets is FIFO-equivalent to the raw per-fill rows (same date → lot order unaffected); notes/exchange union is lossless within now-homogeneous buckets.

Locally: typecheck clean, lint clean, **1294 tests pass** (44 in ibkr.test.ts). 

LGTM — approving.

**Comment by lucasfarre** [comment]
Thanks @GeiserX! 🙌 Really happy to contribute to DeclaRenta — genuinely useful tool. 🙏
